configs/HRTF/alsoft.ini

```diff
@@ -0,0 +1,74 @@
+[general]

```

Comment:
This should probably only include necessary settings, rather than all possible ones. Especially ones like `resampler`, `output-limiter`, and `dither`, which aren't relevant to HRTF.

Reply:
Ohh, I just found out `cubic` (recommended [here](https://youtu.be/62U6UnaUGDE)) was [renamed to `gaussian`](https://github.com/kcat/openal-soft/commit/3ad68aa979de387fa05efad0a081a064c382580b) so it indeed makes sense to exclude it. 👌 

`output-limiter` was set to hopefully address [sounds playing so loud it clips](https://www.youtube.com/watch?v=2ed_mUTmIvs&t=12170s) but even though I haven't noticed that happening after enabling the limiter,  I admit I haven't confirmed whether that really did fix it, or if that that was just a bug with some sounds or DSOAL where a limiter wouldn't help. If there's a better/proper fix, let me know but for now I'll just exclude this option. :

I forgot who/why they recommended `dither` but that one can go too.

And as for the empty flags, I just included them so that the user can easily set values without having to dig up the flags from the template config, but don't they get ignored by the global/default config? 🤔

Reply:
> And as for the empty flags, I just included them so that the user can easily set values without having to dig up the flags from the template config, but don't they get ignored/overridden by the global/default config? 🤔

App-local configs override global ones, and setting a blank value clears whatever value may have been already set. Having `resampler=` in an app-local config file, for example, will cause whatever resampler may be set in a global config to be overridden/ignored and use the default.

The available config options also change with time, with new ones being added and old ones being deprecated or removed, which risks that file getting out of sync if it has unnecessary options. They also lack details and explanations about what the options are and what values can be set, so a user may need to look through the template/sample config file to know what to do with it anyway.

Reply:
Ah I get it now.
How about [commenting out the empty flags](https://github.com/ThreeDeeJay/openal-soft/commit/8865dcdfcd11d61632c3faf5d7dd206347626981)?
![image](https://github.com/kcat/openal-soft/assets/71472458/2aa9a256-d4f2-49d9-ad13-78ee0c8a89e2)


That way users can find flags quickly when instructed to set a specific value.
I also made the workflow [add the config template to the Documentation folder](https://github.com/ThreeDeeJay/openal-soft/commit/e7c0736f7e5524ea655ecca0a510ff7a4a31a23f) for those who want to learn what each flag does. 👌 

![image](https://github.com/kcat/openal-soft/assets/71472458/83235cdd-a880-4625-b4ec-8442e98ea81a)
![image](https://github.com/kcat/openal-soft/assets/71472458/0a95c6ad-1a54-4718-b761-da69e3a2bd43)
![image](https://github.com/kcat/openal-soft/assets/71472458/28830e44-25d7-4718-80b6-9570be890eb6)

Reply:
> Ah I get it now.
> How about [commenting out the empty flags](https://github.com/ThreeDeeJay/openal-soft/commit/8865dcdfcd11d61632c3faf5d7dd206347626981)?

I'd still prefer the unused settings to be removed, to avoid extra maintenance with new or deprecated options. As it is, I'm not sure the config file is even really needed, since WASAPI shouldn't have an issue detecting when the device is set to headphones, and it will just force HRTF when the device doesn't report being headphones. Forcing HRTF when the device isn't headphones, or even if it is headphones but is using virtual surround sound to simulate surround sound speakers, won't sound good and it's better to keep it as speaker output.

FWIW, the `hrtf` option is unneeded (it's deprecated, `stereo-mode=headphones` and `stereo-encoding=hrtf` replace it), and `hrtf-mode=full`, `hrtf-size=0`, and `cf_level=0` are the defaults.

> That way users can find flags quickly when instructed to set a specific value.
> I also made the workflow [add the config template to the Documentation folder](https://github.com/ThreeDeeJay/openal-soft/commit/e7c0736f7e5524ea655ecca0a510ff7a4a31a23f) for those who want to learn what each flag does. 👌

I don't think having unused ones commented out will help much for users being instructed to add settings, since it's not any harder to say to "add `foo=bar` under `[general]`" compared to "uncomment `foo=` and set it to `foo=bar`". Having the template for people who want to read about possible config options is fine, but isn't really helped by having the options commented out in the actual config file. Especially if the real config file has deprecated options that aren't in the template anymore, confusing users, or if the template has options not in the real config file, so they'd have to add it to set it anyway.

Reply:
Fair enough👌 
So this would be the minimal config to ensure HRTF, regardless of values set in the global config, right?
```ini
[general]
channels=stereo
stereo-mode=headphones
stereo-encoding=hrtf
```
tbf I had intentionally kept `hrtf` for backward compatibility in games that require older builds to work around some issues, like [Deus Ex with ALAudio](https://airtable.com/shrtTsUtiAJQ7wFGI/tblNOTdmp5nHXfFGU/viwjDPuKOjBnIQunv/recLK6rvfZs2a40IF) though I think [DSOAL](https://airtable.com/shrtTsUtiAJQ7wFGI/tblNOTdmp5nHXfFGU/viwjDPuKOjBnIQunv/recU6MI6CyiGQOQiW) works better with new builds anyway.

Also, I should point out that the main release package won't have hard-coded config for HRTF, only the template:
https://github.com/ThreeDeeJay/openal-soft/releases/download/latest/OpenALSoft.zip 
https://github.com/ThreeDeeJay/openal-soft/releases/download/latest/OpenALSoft+HRTF.zip as an separate file should make it obvious that the intent is to use HRTF, even if the device's set to surround (because I think the average person just uses the onboard sound and might not even realize if the system config is set to surround, so not forcing HRTF might actually cause missing audio channels on since only front left and front right would be heard on headphones)

> WASAPI shouldn't have an issue detecting when the device is set to headphones

wait, so is it possible to change the windows audio device type or do you mean via `stereo-mode`?
IME OpenAL Soft seems to always detect surround-capable devices as speakers (even when set to stereo), and only certain devices like the front panel, DACs (and ofc headphones with built-in DACs) as headphones.

> if it is headphones but is using virtual surround sound to simulate surround sound speakers, won't sound good 

Yeah, unfortunately that's one of the inconveniences we have to deal with. 
The only possible solutions I could think of are https://github.com/kcat/openal-soft/issues/586, https://github.com/kcat/openal-soft/issues/682 or somehow interfacing with a VSS like HeSuVi via EqualizerAPO, or replacing them altogether with a custom APO, though I guess that's a bit outside of the project's scope. 🤔

Reply:
Sorry, I somehow missed replying.

> So this would be the minimal config to ensure HRTF, regardless of values set in the global config, right?

Yes.

> wait, so is it possible to change the windows audio device type or do you mean via `stereo-mode`?

I mean in Windows. It's been a while since I dealt with it, I don't use Windows often and I don't have access to a Windows system right now, but I remember there being a dropdown selection somewhere in the audio device configuration to say whether a stereo device is Speakers or Headphones, which correlated to the FormFactor property that OpenAL Soft's WASAPI backend reads to autodetect headphones.

If you're not seeing it, maybe it's only available for certain devices/drivers, or it was moved/removed in other versions of Windows, which would be unfortunate.

Reply:
No worries. And I'm not aware of a setting to set device type on the Windows GUI side, usually only on the driver's control panel, but that doesn't seem to change how Windows itself treats the device. Like on my USB X-Fi, it just enables Headphone HRTF when CMSS-3D is enabled, as opposed to speaker HRTF otherwise IIRC, but Windows still treats it as a speaker.
![image](https://github.com/user-attachments/assets/93525c5f-4925-4054-b916-ada451588938)
AFAIK the headphones and stereo speaker configurations were merged into just Stereo after [Windows XP](https://developer.valvesoftware.com/wiki/Talk:Snd_surround_speakers). 
On a side note, have you considered letting some [games](https://i.imgur.com/cqGjf3Z.png)/[apps](https://imgur.com/lT2xGzh)' headphone setting override the system's when using DSOAL in case it can be changed?

Though I suspect the device type may be buried in a [registry device property key](https://github.com/AutomatedLab/DeviceManagement/blob/master/DeviceManagementLib/Device.cs#L113-L289) like `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render\DEVICE_GUID\Properties`, unless it's hard coded into the driver. Here's the registry values from my onboard Realtek front (Headphones) vs rear (Speakers) panel playback devices: [RealtekRegistryProperties.zip](https://github.com/user-attachments/files/16503195/RealtekRegistryProperties.zip)
![image](https://github.com/user-attachments/assets/7cb639ce-ffad-4a61-b4e5-7086a508d01c)

My X-Fi only has a single headphone port, though I think it used to automatically switch to headphones when it was plugged in but I think that only work on an older Windows/driver. But I digress.

@mirh Are you familiar with the playback device type (headphones vs speaker) property on modern Windows?

Reply:
> On a side note, have you considered letting some [games](https://i.imgur.com/cqGjf3Z.png)/[apps](https://imgur.com/lT2xGzh)' headphone setting override the system's when using DSOAL in case it can be changed?

Maybe. As far as I know, the app controls that with `IDirectSound::SetSpeakerConfig` (and `IDirectSound::GetSpeakerConfig`). They're no-ops since Windows Vista according to Wine, likely to prevent individual apps from changing the system's device configuration, which is what DSOAL does too. Theoretically I could maybe watch for it to be set to `DSSPEAKER_HEADPHONE` and enable HRTF for the device, and leave it at the default for everything else. I'd have a slight worry about it potentially confusing apps since the setting wouldn't persist between closing/reopening the device, though an app probably couldn't rely on it persisting anyway since other apps could change it too, so it may not be a big deal.

> @kcat By the way, is there anything else I should change on this PR? Besides resolving merge conflicts, of course 😅

No, I think it looks alright once the conflicts are fixed.

Reply:
You can't really expect creative's drivers to be representative of how things should properly work 🙃
(and regardless, I remember even my Xonar has an "internal separation" between what it presents to the system and the one it knows is physically connected to the the card)
And we had already discussed speakers and headphones in https://github.com/kcat/dsoal/issues/63 (applications shouldn't interfere with system configuration, but of course there's always edge cases). 

Nevertheless the property is there underneath (even though I could swear there was also somewhere a dropdown menu that let you select it, as kcat said.. but I cannot find it now)

Reply:
> No, I think it looks alright once the conflicts are fixed.

@kcat Done. And while we're at it, let me know if https://github.com/kcat/dsoal/pull/52 needs any more changes 👀 
I'm planning to add a separate DSOAL build with HRTF too, when I can fetch the config from OpenAL Soft upstream, but it should be good to go as it is now. 👌 

> I could swear there was also somewhere a dropdown menu that let you select it, as kcat said.. but I cannot find it now

@mirh Are you referring to this?

![firefox_q2GAprN2UQ](https://github.com/user-attachments/assets/ef13e7c6-89b2-4855-8de6-bc57b0a318c0)
Apparently that dialog only shows up for the front panel which is already detected as headphones, and it doesn't even respect the speaker mode

Reply:
That, I still get it. 
And that includes the back panel too. 

But I just could remember this native windows control panel with the selection...

Reply:
like this flyout dialog or like something deeper in the sound settings?
![image](https://github.com/user-attachments/assets/9722f490-e4ab-4db2-a9da-3d57f2a20074)

> That, I still get it.
And that includes the back panel too.

Do any of them change how Windows detects the device?
I wonder if some sound cards just switch like a driver mode to present/cloak itself as a "different" speakers or headphones device instead of telling Windows to change it on its end. 🤔